src/pcre2_compile.c

```diff
@@ -4151,20 +4269,106 @@ while (ptr < ptrend)
 
+      /* Forbid unescaped literals, and the special meaning of '-', inside a
+      Perl extended class. */
+
+      else if (class_mode_state == CLASS_MODE_PERL_EXT)
+        {
+        errorcode = ERR116;
+        goto FAILED;
+        }
+
+      /* Handle potential start of range */
+
+      else if (c == CHAR_MINUS && class_range_state >= RANGE_OK_ESCAPED)
+        {
+        *parsed_pattern++ = (class_range_state == RANGE_OK_LITERAL)?
+          META_RANGE_LITERAL : META_RANGE_ESCAPED;
+        class_range_state = RANGE_STARTED;
+        }
+
+      /* Handle forbidden start of range */
+
+      else if (c == CHAR_MINUS && class_range_state == RANGE_FORBID_NO)
+        {
+        *parsed_pattern++ = CHAR_MINUS;
+        class_range_state = RANGE_FORBID_STARTED;
+        class_range_forbid_ptr = ptr;
+        }
+
+      /* Handle a literal character */
+
+      else
+        {
+        CLASS_LITERAL:
+
+        /* Disallow implicit union in Perl extended classes. */
+
+        if (class_op_state == CLASS_OP_OPERAND &&
+            class_mode_state == CLASS_MODE_PERL_EXT)
+          {
+          errorcode = ERR113;
+          goto FAILED;
+          }
+
+        if (class_range_state == RANGE_STARTED)
+          {
+          if (c == parsed_pattern[-2])       /* Optimize one-char range */
+            parsed_pattern--;
+          else if (parsed_pattern[-2] > c)   /* Check range is in order */
+            {
+            errorcode = ERR8;
+            goto FAILED_BACK;  // TODO https://github.com/PCRE2Project/pcre2/issues/549
+            }
+          else
+            {
+            if (!char_is_literal && parsed_pattern[-1] == META_RANGE_LITERAL)
+              parsed_pattern[-1] = META_RANGE_ESCAPED;
+            PARSED_LITERAL(c, parsed_pattern);
+            }
+          class_range_state = RANGE_NO;
+          class_op_state = CLASS_OP_OPERAND;
+          }
+        else if (class_range_state == RANGE_FORBID_STARTED)
+          {
+          ptr = class_range_forbid_ptr;
+          errorcode = ERR50;
+          goto FAILED;
+          }
+        else  /* Potential start of range */
+          {
+          class_range_state = char_is_literal?
+            RANGE_OK_LITERAL : RANGE_OK_ESCAPED;
+          class_op_state = CLASS_OP_OPERAND;
+          PARSED_LITERAL(c, parsed_pattern);
+          }
+        }
+
       /* Proceed to next thing in the class. */
 
       CLASS_CONTINUE:
       if (ptr >= ptrend)
         {
-        if ((options & PCRE2_ALT_EXTENDED_CLASS) != 0 &&
+        if (class_mode_state == CLASS_MODE_PERL_EXT && class_depth_m1 > 0)
+          errorcode = ERR14;   /* Missing terminating ')' */
+        if (class_mode_state == CLASS_MODE_ALT_EXT &&
             class_depth_m1 == 0 && class_maxdepth_m1 == 1)
           errorcode = ERR112;  /* Missing terminating ']', but we saw '[ [ ]...' */
         else
-          errorcode = ERR6;  /* Missing terminating ']' */
+          errorcode = ERR6;    /* Missing terminating ']' */
```

Comment:
Why these spaces are added?

Reply:
So the three comments `/* Missing terminating ...` line up nicely :)

There are lots of examples of vertically-aligned comments and formatting in PCRE2, which I like too.